File: benchmarks/runners/python/emit.py
"""Collector/main: run cold-start/first-run (fresh processes) + warm/streaming
(in-process), validate ids, stamp env, write results/python-<ts>.json. Emit-only —
does NOT render the report (root :benchmarkCompare renders once over all runners)."""

import json
from datetime import datetime, timezone
from pathlib import Path

from coldstart import run_cold_start_and_first_run
from env import gather_env
from manifest import load_manifest, validate_result_ids
from warm_bench import run_warm_and_streaming
from wrapper import load_wrapper

# benchmarks/runners/python -> benchmarks
_BENCH_DIR = Path(__file__).resolve().parents[2]
CORPUS = _BENCH_DIR / "corpus"
RESULTS_DIR = _BENCH_DIR / "results"


def build_result(env, cases):
    ts = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z"
    return {
        "schemaVersion": "1.0",
        "runner": env["runner"],
        "env": env,
        "timestamp": ts,
        "cases": cases,
    }


def main():
    manifest = load_manifest(CORPUS)

    # Cold-start / first-run first (fresh processes), then warm/streaming in-process.
    cold_rows = run_cold_start_and_first_run(manifest)

    api = load_wrapper()
    dw = api.DataWeave()
    dw.initialize()
    try:
        warm_rows = run_warm_and_streaming(dw, manifest)
    finally:
        dw.cleanup()

    cases = cold_rows + warm_rows
    validate_result_ids(manifest, [c["id"] for c in cases])  # fail-fast on orphan ids

    env = gather_env()
    result = build_result(env, cases)

    RESULTS_DIR.mkdir(parents=True, exist_ok=True)
    stamp = result["timestamp"].replace(":", "-").replace(".", "-")
    out = RESULTS_DIR / f"python-{stamp}.json"
    out.write_text(json.dumps(result, indent=2), encoding="utf-8")
    print(f"wrote {out} ({len(cases)} rows)")
    return out


if __name__ == "__main__":
    main()

File: benchmarks/runners/node/emit.mjs
import { writeFileSync, mkdirSync } from "node:fs";
import { join, dirname } from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";
import { loadManifest, validateResultIds } from "../../lib/manifest.mjs";
import { gatherEnv } from "../../lib/env.mjs";
import { loadWrapper } from "./wrapper.mjs";
import { runWarmAndStreaming } from "./warm-bench.mjs";
import { runColdStartAndFirstRun } from "./coldstart.mjs";

const __dirname = dirname(fileURLToPath(import.meta.url));
const CORPUS = join(__dirname, "..", "..", "corpus");
const RESULTS_DIR = join(__dirname, "..", "..", "results");

/** Assemble the full schema object. */
export function buildResult(env, cases) {
  return {
    schemaVersion: "1.0",
    runner: env.runner,
    env,
    timestamp: new Date().toISOString(),
    cases,
  };
}

export async function main() {
  const manifest = loadManifest(CORPUS);
  const env = gatherEnv({ runner: "node-wrapper", runtimeVersion: `node ${process.version}` });

  // Cold-start / first-run first (fresh processes), then warm/streaming in-process.
  const coldRows = await runColdStartAndFirstRun(manifest);
  const api = await loadWrapper();
  let warmRows;
  try {
    warmRows = await runWarmAndStreaming(api, manifest);
  } finally {
    api.cleanup();
  }

  const cases = [...coldRows, ...warmRows];
  validateResultIds(manifest, cases); // fail-fast on any orphan id

  mkdirSync(RESULTS_DIR, { recursive: true });
  const stamp = new Date().toISOString().replace(/[:.]/g, "-");
  const outPath = join(RESULTS_DIR, `node-${stamp}.json`);
  writeFileSync(outPath, JSON.stringify(buildResult(env, cases), null, 2));
  console.log(`wrote ${outPath} (${cases.length} rows)`);
  return outPath;
}

// Run when invoked directly.
if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  main().catch((e) => {
    console.error(e.message);
    process.exit(1);
  });
}
